src/machines/java/com/enderio/machines/common/io/fluid/MachineFluidTank.java

```diff
@@ -0,0 +1,244 @@
+        super(capacity, validator);
+        this.parent = parent;
+    }
+
+    /** This implementation only handles one tank, use multiple instances of this class in combination
+     *      with a FluidTankHandler to use multiple tanks
+     * @return - 1.
+     */
+    @Override
+    public int getTanks() {
+        return 1;
+    }
+
+    /** Transfers fluid from a fluid handler to this tank.
+     * @param from - fluid handler to transfer from
+     * @param desiredAmount - The max amount to transfer if possible.
+     * @param force - Force transfer even if 'allowInput' is false or validator prohibits fluid to enter. It respects the fluid types though
+     * @return - The amount that was actually transferred.
+     */
+    public int transferFrom(IFluidHandler from, int desiredAmount, boolean force) {
+        FluidStack incomingStack = from.drain(desiredAmount, FluidAction.SIMULATE);
+        int transferredAmount = fill(incomingStack, FluidAction.EXECUTE,force);
+        from.drain(transferredAmount, FluidAction.EXECUTE);
+        return transferredAmount;
+    }
+
+    /**
+     * @param to - The handler to transfer into.
+     * @param desiredAmount - The max amount to transfer if possible.
+     * @param force - Force transfer even if 'allowOutput' is false.
+     * @return - The amount that was actually transferred.
+     */
+    public int transferTo(IFluidHandler to, int desiredAmount, boolean force){
+        FluidStack outStack = fluid.copy();
+        int transferAmount = 0;
+        if(!outStack.isEmpty()){
+            outStack.setAmount(drain(desiredAmount,FluidAction.SIMULATE, force));//caps at the tank's capacity to transfer out of and at the desired amount
+            transferAmount = to.fill(outStack, FluidAction.EXECUTE);
+        }
+        drain(transferAmount, FluidAction.EXECUTE);//should always succeed to transfer what we have simulated
+        return transferAmount;
+    }
+
+
+
+    /** Works like the FluidTank class of forge with some additions described below.
+     *
+     * @param source - The source fluid stack. Tries to increase tank amount with entire source amount.
+     *               WARNING! This method does not drain the source stack according to the forge FluidTank implementation.
+     *               For that use case, use the method "transferFrom"
+     * @param action - Execute makes the transfer happen, simulate checks what would happen if executed.
+     * @param force - Force your way around the allow input flag and the validator (filter). Usually used to
+     *              fill a machine output tank from within the machine.
+     * @return - The amount that could be filled in case of simulation or was filled in case of execution.
+     *
+     * Credit: the Forge team, since the logic is their work.
+     */
+    public int fill(FluidStack source, FluidAction action, boolean force){
+        if (source.isEmpty() || (!fluid.isEmpty() && !fluid.isFluidEqual(source))) return 0;
```

Comment:
use pattern like this:
```
if (source.isEmpty)
    return 0;
if (!fluid.isEmpty() && !fluid.isFluidEqual(source))
    return 0;
```

Reply:
microoptimization like that are probably not worth it, if they make code less readable and I kinda doubt that fill is called often enough to be worth it

Reply:
Okay!